articles/tuples-data-structure-python/index.md

```diff
@@ -0,0 +1,307 @@
+```
+
+This is because a tuple is designed to be immutable. However, we can change a tuple that contains mutable objects. For example, let us take a tuple of lists.
+
+```py
+tuple1 = ([10, 20], [30, 40], [50, 60])
+
+tuple1[1][0] = 70
+
+print(tuple1) # Output: ([10, 20], [70, 40], [50, 60])
+```
+
+This works perfectly because we are modifying the list within a tuple (which is mutable). We can also create new tuples from existing ones.
+
+```py
+tuple1 = ([10, 20], [30, 40], [50, 60])
+tuple2 = ([100, 200], [300, 400])
+
+# Creating a new tuple from tuple1 and tuple2
+tuple3 = tuple1 + tuple2
+
+print(tuple3) # Output: ([10, 20], [30, 40], [50, 60], [100, 200], [300, 400])
+```
+
+#### Tuple Methods
+
+Tuples have the following in-built methods that make them extremely powerful:
+
+- **`cmp(tuple1, tuple2)`**
+- **`len(tuple)`**
+- **`min(tuple)`**
+- **`max(tuple)`**
+- **`tuple(list)`**
+- **`t.count(el)`**
+- **`t.index(el)`**
+
+##### cmp(tuple1, tuple2)
+
+_Note: The cmp() method existed in python2. It wasn't included in python3. Therefore we define our own compare method._
+
+The compare method analyses two tuples element by element. It compares them and returns the following:
+
+- If tuple1 > tuple2: the method returns 1.
+- If tuple2 > tuple1: the method returns -1.
+- If tuple1 == tuple2: the method returns 0.
+
+```py
+def cmp(t1, t2):
+    return bool(t1 > t2) - bool(t1 < t2)
+    """
+    When both tuples are equal,
+        bool(t1 > t2) = 0
+        bool(t1 < t2) = 0
+        Therefore, 0 - 0 = 0
+
+    When both tuple1 > tuple2,
+        bool(t1 > t2) = 1
+        bool(t1 < t2) = 0
+        Therefore, 1 - 0 = 0
+
+    When both tuple2 > tuple1,
+        bool(t1 > t2) = 0
+        bool(t1 < t2) = 1
+        Therefore, 0 - 1 = -1
+    """
+tuple1 = (100, 200)
+tuple2 = (300, 400)
+
+print(cmp(tuple1, tuple2))
+# Output: -1
+
+print(cmp(tuple2, tuple1))
+# Output: 1
+
+tuple1 = (100, 200)
+tuple2 = (100, 200)
+
+print(cmp(tuple2, tuple1))
+# Output: 0
+```
+
+##### len(tuple)
+
+The length method returns the length of the tuple.
+
+```py
+tuple1 = (10, 20, 30, 40, 50)
+
+print(len(tuple1))
+# Output: 5
+```
+
+##### min(tuple)
+
+The min method returns the smallest element in the tuple.
+
+```py
+tuple1 = (3, 9, 1, 90, 200)
+
+print(min(tuple1))
+# Output: 1
+```
+
+##### max(tuple)
+
+The max method returns the largest element in the tuple.
+
+```py
+tuple1 = (3, 9, 1, 90, 200)
+
+print(max(tuple1))
+# Output: 200
+```
+
+##### tuple(list)
+
+The tuple method converts the list that is passed as parameter into a tuple.
+
+```py
+list1 = [23, 34, 45, 56]
+
+print(tuple(list1))
+# Output: (23, 34, 45, 56)
+```
+
+##### t.count(el)
+
+The count method returns the count of the element passed as parameter.
+
+```py
+tuple1 = (1, 24, 45, 54, 6, 34, 24)
+
+print(tuple1.count(24))
+# Output: 2
+```
+
+##### t.index(el)
+
+The index method returns the index of the first occurence of the element in a tuple.
```

Comment:
is there a way to get the last occurrence of the element in a tuple? Check this [out](https://www.w3resource.com/python-exercises/tuple/python-tuple-exercise-14.php).

Reply:
Added some examples for getting the last occurrence, and also to specify a range to search from.